File: 1423-maximum-number-of-occurrences-of-a-substring/maximum-number-of-occurrences-of-a-substring.java
import java.util.*;

class Solution {
    public int maxFreq(String s, int maxLetters, int minSize, int maxSize) {
        Map<String, Integer> freq = new HashMap<>();
        int res = 0;

        for (int i = 0; i <= s.length() - minSize; i++) {
            String sub = s.substring(i, i + minSize);

            if (countDistinct(sub) <= maxLetters) {
                freq.put(sub, freq.getOrDefault(sub, 0) + 1);
                res = Math.max(res, freq.get(sub));
            }
        }
        return res;
    }

    private int countDistinct(String str) {
        Set<Character> set = new HashSet<>();
        for (char c : str.toCharArray()) {
            set.add(c);
        }
        return set.size();
    }
}

File: 1423-maximum-number-of-occurrences-of-a-substring/maximum-number-of-occurrences-of-a-substring.py
class Solution:
    def maxFreq(self, s: str, maxLetters: int, minSize: int, maxSize: int) -> int:
        freq = {}
        res = 0 

        for i in range(len(s) - minSize + 1):
            sub = s[i : i + minSize]

            if (len(set(sub))) <= maxLetters:
                freq[sub] = freq.get(sub , 0) + 1

                res = max(res , freq[sub])
        return res
